service/interpreter/timers/greedyTimerProcessor.go

```diff
@@ -0,0 +1,264 @@
+	tp.createGreedyTimerScheduler(ctx, continueAsNewCounter)
+
+	err := provider.SetQueryHandler(ctx, service.GetCurrentTimerInfosQueryType, func() (service.GetCurrentTimerInfosQueryResponse, error) {
+		return service.GetCurrentTimerInfosQueryResponse{
+			StateExecutionCurrentTimerInfos: tp.stateExecutionCurrentTimerInfos,
+		}, nil
+	})
+	if err != nil {
+		panic("cannot set query handler")
+	}
+	return tp
+}
+
+func (t *sortedTimers) addTimer(toAdd *service.TimerInfo) {
+
+	if toAdd == nil || toAdd.Status != t.status {
+		panic("invalid timer added")
+	}
+
+	insertIndex := 0
+	for i, timer := range t.timers {
+		if toAdd.FiringUnixTimestampSeconds >= timer.FiringUnixTimestampSeconds {
+			// don't want dupes. Makes remove simpler
+			if toAdd == timer {
+				return
+			}
+			insertIndex = i
+			break
+		}
+		insertIndex = i + 1
+	}
+	t.timers = append(
+		t.timers[:insertIndex],
+		append([]*service.TimerInfo{toAdd}, t.timers[insertIndex:]...)...)
+}
+
+func (t *sortedTimers) removeTimer(toRemove *service.TimerInfo) {
+	for i, timer := range t.timers {
+		if toRemove == timer {
+			t.timers = append(t.timers[:i], t.timers[i+1:]...)
+			return
+		}
+	}
+}
+
+func (t *sortedTimers) pruneToNextTimer(pruneTo int64) *service.TimerInfo {
+
+	if len(t.timers) == 0 {
+		return nil
+	}
+
+	index := len(t.timers)
+
+	for i := len(t.timers) - 1; i >= 0; i-- {
+		timer := t.timers[i]
+		if timer.FiringUnixTimestampSeconds > pruneTo && timer.Status == t.status {
+			break
+		}
+		index = i
+	}
+	t.timers = t.timers[:index]
+	return t.timers[index-1]
+}
+
+func (t *GreedyTimerProcessor) createGreedyTimerScheduler(
+	ctx interfaces.UnifiedContext,
+	continueAsNewCounter *cont.ContinueAsNewCounter) {
+
+	t.provider.GoNamed(ctx, "greedy-timer-scheduler", func(ctx interfaces.UnifiedContext) {
+		// NOTE: next timer to fire is at the end of the slice
+		var createdTimers []int64
+		for {
+			t.provider.Await(ctx, func() bool {
+				// remove fired timers
+				now := t.provider.Now(ctx).Unix()
+				for i := len(createdTimers) - 1; i >= 0; i-- {
+					if createdTimers[i] > now {
+						createdTimers = createdTimers[:i+1]
+						break
+					}
+				}
+				next := t.pendingTimers.pruneToNextTimer(now)
+				return (next != nil && (len(createdTimers) == 0 || next.FiringUnixTimestampSeconds < createdTimers[len(createdTimers)-1])) || continueAsNewCounter.IsThresholdMet()
+			})
+
+			if continueAsNewCounter.IsThresholdMet() {
+				break
+			}
+
+			now := t.provider.Now(ctx).Unix()
+			next := t.pendingTimers.pruneToNextTimer(now)
+			//next := t.pendingTimers.getEarliestTimer()
+			// only create a new timer when a pending timer exists before the next existing timer fires
+			if next != nil && (len(createdTimers) == 0 || next.FiringUnixTimestampSeconds < createdTimers[len(createdTimers)-1]) {
+				fireAt := next.FiringUnixTimestampSeconds
+				duration := time.Duration(fireAt-now) * time.Second
+				t.provider.NewTimer(ctx, duration)
```

Comment:
This NewTimer is a non-blocking one. 
At first glance I thought it should be `t.provider.Sleep(...)` then I remember that we had a similar discussion when implemeting it for Wayfinder LoL.

So we don't need to wait here because we rely on the fact that when the timer fired, workflow task will get triggered and evaluate the Await condition, right? 

Maybe add more comments here too.